modal/_container_io_manager.py

```diff
@@ -146,22 +148,20 @@ async def _heartbeat_handle_cancellations(self) -> bool:
         # Return True if a cancellation event was received, in that case
         # we shouldn't wait too long for another heartbeat
 
-        # Don't send heartbeats for tasks waiting to be checkpointed.
-        # Calling gRPC methods open new connections which block the
-        # checkpointing process.
-        if self._waiting_for_memory_snapshot:
-            return False
-
         request = api_pb2.ContainerHeartbeatRequest(supports_graceful_input_cancellation=True)
         if self.current_input_id is not None:
             request.current_input_id = self.current_input_id
         if self.current_input_started_at is not None:
             request.current_input_started_at = self.current_input_started_at
 
-        # TODO(erikbern): capture exceptions?
-        response = await retry_transient_errors(
-            self._client.stub.ContainerHeartbeat, request, attempt_timeout=HEARTBEAT_TIMEOUT
-        )
+        async with self._pause_heartbeats:
+            while self._waiting_for_memory_snapshot:
```

Comment:
Maybe dumb q but I can't see why this is a `while` and not just an `if`. If this is obvious, could it get a comment so I don't hurt myself thinking next time I come across the code 😄

Reply:
Posted answer to this in thread above by accident